packages/react-i18n/src/hooks.tsx

```diff
@@ -49,8 +53,30 @@ function useComplexI18n(
   // reasons, it's safe to just store the IDs once and never let them change.
   const ids = useLazyRef(() => (id ? [id, ...parentIds] : parentIds));
 
-  if (id && (translations || fallback)) {
-    manager.register({id, translations, fallback});
+  // When the manager changes, we need to do the following IMMEDIATELY (i.e.,
+  // not in a useEffect callback):
+  //
+  // 1. Register the component’s translations. This ensures that the first render gets
+  //    the synchronous translations, if available.
+  // 2. Unsubscribe from changes to a previous manager.
+  // 3. Subscribe to changes from the new manager. This ensures that if the subscription
+  //    is updated between render and `useEffect`, the state update is not lost.
+  if (manager !== managerRef.current) {
```

Comment:
Is there any way we can add a regression test for this?

Reply:
I can't think of how to test the actual issue this PR addresses. I could definitely test the "manager changing" bit, but I don't think that's really a common case to worry about. The problem here is one of timing, and I am not really sure how I can force an async translation to be marked as complete (and have the resulting enqueued update happen) before `useEffect` is called.